[SelectionDAG] Add an ISD node for for get.active.lane.mask

kmclaughlin-arm opened the pull request:
For now expansion still happens in SelectionDAGBuilder when
GET_ACTIVE_LANE_MASK is not legal on the target.

This patch also includes changes in AArch64ISelLowering to replace
handling of the get.active.lane.mask intrinsic to use the ISD node.
Tablegen patterns are added which match to whilelo for scalable types.

A follow up change will add support for more types to be lowered to
GET_ACTIVE_LANE_MASK by allowing splitting of the node.

kmclaughlin-arm committed:
[SelectionDAG] Add an ISD node for for get.active.lane.mask

For now expansion still happens in SelectionDAGBuilder when
GET_ACTIVE_LANE_MASK is not legal on the target.

This patch also includes changes in AArch64ISelLowering to replace
handling of the get.active.lane.mask intrinsic to use the ISD node.
Tablegen patterns are added which match to whilelo for scalable types.

A follow up change will add support for more types to be lowered to
GET_ACTIVE_LANE_MASK by allowing splitting of the node.

paulwalker-arm commented:
Given many of the changes are required to maintain existing code quality after loosing the lowering to the `whilelo` intrinsic I think it's worth seeing if you can also change `performIntrinsicCombine` to always lower the intrinsic itself to `GET_ACTIVE_LANE_MASK`. That way all the code paths will be unified and we'll not need extract isel patterns because the `GET_ACTIVE_LANE_MASK` can replace the existing intrinsic ones?

kmclaughlin-arm committed:
- Lower whilelo intrinsic to GET_ACTIVE_LANE_MASK in performIntrinsicCombine
- Remove the additional tablegen patterns matching get_active_lane_mask
  and reuse those for the whilelo intrinsic

kmclaughlin-arm committed:
- Move isBeforeLegalize before call to optimizeIncrementingWhile
- Add isSVEorStreamingSVEAvailable assert to LowerGET_ACTIVE_LANE_MASK
- Added more details to the comment in ISDOpcodes.h

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -29515,6 +29509,30 @@ AArch64TargetLowering::LowerPARTIAL_REDUCE_MLA(SDValue Op,
+  EVT VT = Op.getValueType();
+  assert(VT.isFixedLengthVector() && "Expected fixed length vector type!");
+
+  auto &Subtarget = DAG.getSubtarget<AArch64Subtarget>();
```
This will generate an "unsused variable" compiler warning for release builds.  It is best to have everything in the assert and in this case I think `AArch64TargetLowering` already has a `Subtarget` member you can use?

kmclaughlin-arm committed:
- Use Subtarget from AArch64TargetLowering in LowerGET_ACTIVE_LANE_MASK
